tests/test_dtypes.py

```diff
@@ -1,436 +0,0 @@
-            return y + x
-        elif x.dtype == Unsigned96() and y.dtype == ndx.uint32:
-            return x + y.astype(Unsigned96())
-        return NotImplemented
-
-    def where(self, condition, x, y):
-        x = x.astype(Unsigned96())
-        y = y.astype(Unsigned96())
-        return super().where(condition, x, y)
-
-
-class Unsigned96(StructType, CastMixin):
-    def _fields(self) -> dict[str, StructType | CoreType]:
-        return {
-            "upper": ndx.uint64,
-            "lower": ndx.uint32,
-        }
-
-    def _parse_input(self, x: np.ndarray) -> dict:
-        upper = self._fields()["upper"]._parse_input(
-            np.array(x >> 32).astype(np.uint64)
-        )
-        lower = self._fields()["lower"]._parse_input(
-            (x & np.array([0xFFFFFFFF])).astype(np.uint32)
-        )
-        return {
-            "upper": upper,
-            "lower": lower,
-        }
-
-    def _assemble_output(self, fields: dict[str, np.ndarray]) -> np.ndarray:
-        return (fields["upper"].astype(object) << 32) | fields["lower"].astype(object)
-
-    def copy(self) -> Self:
-        return self
-
-    def _schema(self) -> Schema:
-        return Schema(type_name="u96", author="value from data!")
-
-    def _cast_to(self, array: Array, dtype: CoreType | StructType) -> Array:
-        raise CastError(f"Cannot cast {self} to {dtype}")
-
-    def _cast_from(self, array: Array) -> Array:
-        if array.dtype == ndx.uint32:
-            array = array.astype(ndx.uint32)
-        else:
-            raise CastError(f"Cannot cast {array.dtype} to {self}")
-
-        return Array._from_fields(
-            Unsigned96(), upper=ndx.asarray(0, dtype=ndx.uint64), lower=array
-        )
-
-    _ops = Unsigned96Impl()
-
-
-class ListImpl(OperationsBlock):
-    def make_array(
-        self,
-        shape: tuple[int | str | None, ...],
-        dtype: CoreType | StructType,
-        eager_value: np.ndarray | None = None,
-    ) -> Array:
-        if eager_value is None:
-            return Array._from_fields(
-                dtype,
-                endpoints=ndx.array(shape=shape + (2,), dtype=ndx.int64),
-                items=ndx.array(shape=(None,), dtype=ndx.utf8),
-            )
-        else:
-            fields = dtype._parse_input(eager_value)
-            return Array._from_fields(
-                dtype, **{name: ndx.asarray(field) for name, field in fields.items()}
-            )
-
-    def getitem(
-        self,
-        x: Array,
-        index,
-    ) -> Array:
-        if isinstance(index, int):
-            index = slice(index, index + 1), ...
-
-        return Array._from_fields(
-            dtype=x.dtype,
-            endpoints=x.endpoints[index],
-            items=x.items.copy(),
-        )
-
-    def shape(self, x) -> Array:
-        return nda.shape(x.endpoints)[:-1]
-
-    def static_shape(self, x) -> tuple[int | None, ...]:
-        return x.endpoints.shape[:-1]
-
-
-class List(StructType):
```

Comment:
This test was very valuable for testing the limits of the dtype system. I'm sure it's implementable again - can you please bring back the `List` test and/or have some custom dtype tests that illustrate how dtypes get defined from outside of ndonnx. Right now, everything (categoricals, object dtype) has been inlined into the library which won't be the long term way to support niche dtypes without a clear numpy counterpart.